[WIP] Check for overflow on integer arithmetic

chriseth opened the pull request:
related to https://github.com/ethereum/solidity/issues/2274

This is, mostly, a starting place and a very rough and very early
work-in-progress for detecting possible arithmetic overflow conditions.
Right now, it works for basic arithmetic and assignment, does not
analyze branches or control structures, and probably works only for
one function in one contract.

roadriverrail committed:
Check for overflow on integer arithmetic

This is, mostly, a starting place and a very rough and very early
work-in-progress for detecting possible arithmetic overflow conditions.

Currently supported is: basic arithmetic, assignment, basic if/else
(with < and > operators only), multiple functions per contract, calling
functions

Most strongly needed: fix constraints so that == and != are supported,
while and for loops, warning if divide by 0 is possible

chriseth commented:
Thanks for the pull request!

I think that in general, it would probably make sense to separate the "range" and the "allowed range" aspect, also looking towards a more generalized use of these classes. Furthermore, what happens in a binary operator (unless it is a comparison operator) is that the operands are first converted into the common type and then the operation is performed in that common type. Since these are implicit conversions, the range will always fit, but the overflow will happen in the new type. Also, I think it would be beneficial to have a bounds constructor taking a type object (instead of "number of bytes" and "signed"). This means, the operand visitor would be something like:

```
Bounds b = combineBounds(operand, leftBounds, rightBounds);
if (!b.fits(Bounds(commonType)))
  warning(...).
```

Ah and also note that variables can be used before they are declared, so this basically means that whenever `m_varOverflowInfo` does not contain `&var`, just initialize it with the default range of `var`'s type and you do not have to care about `visit(VariableDeclaration...)` at all.

axic commented on libsolidity/interface/CompilerStack.cpp:
```diff
@@ -215,6 +216,8 @@ bool CompilerStack::analyze()
 				{
 					contract->setDevDocumentation(Natspec::devDocumentation(*contract));
 					contract->setUserDocumentation(Natspec::userDocumentation(*contract));
```
This is not strictly related to this PR, but why is the documentation stuff in this conditional?

roadriverrail replied:
I don't know the first thing about documentation, so I am not qualified to answer your question.

chriseth commented on libdevcore/Common.h:
```diff
@@ -101,6 +101,16 @@ using u160s = std::vector<u160>;
 using u160Set = std::set<u160>;
 
+// Numeric types with checked arithmetic
+using safe_u64 =  boost::multiprecision::number<boost::multiprecision::cpp_int_backend<64, 64, boost::multiprecision::unsigned_magnitude, boost::multiprecision::checked, void>>;
```
These are not really required, are they?

roadriverrail replied:
No.  That's dead code.  I'm removing it in a cleanup.

chriseth commented on libsolidity/analysis/OverflowBounds.h:
```diff
@@ -0,0 +1,59 @@
+namespace solidity
+{
+
+class OverflowBounds
```
Is the word "Overflow" justified here? What about something like "ValueRanges" (we might generalize this to something not just about bounds).

roadriverrail replied:
No, probably not.  I had wanted to make a generic class that would be useful for both reasoning about integers and booleans.  The more I work with it, however, the worse that idea even seems.  I am starting to feel like I should just get rid of the parent class until I know what I'm doing with it anyway.  What do you think?

chriseth commented on libsolidity/analysis/IntegerOverflowBounds.cpp:
```diff
@@ -0,0 +1,267 @@
+	solAssert((lbounds && rbounds), "Passed non-integer bounds objects to binary op for an integer bounds");
+	performBinaryOp(op, *lbounds, *rbounds);
+}
+	void IntegerOverflowBounds::performBinaryOp(
```
I somehow expected this function to be static and returning a new `IntegerOverflowBounds`. Why is that not the case?

roadriverrail replied:
There's no strong reason for it.  It'll probably make for cleaner code in its consumer.  Will rewrite.

chriseth commented on libsolidity/analysis/IntegerOverflowBounds.cpp:
```diff
@@ -0,0 +1,267 @@
+	if (isSigned)
+	{
+		allowedMin = 1;
+		allowedMin = allowedMin << exponent;
```
Perhaps there should be a separate file that has free functions that return the min / max values for each type?

roadriverrail replied:
I'm not against it.  I want to do a code reorg before I proceed.  When you say "each type," do you mean IntegerType and RationalNumberType?

chriseth commented on libsolidity/analysis/IntegerOverflowBounds.cpp:
```diff
@@ -0,0 +1,267 @@
+		allowedMax = allowedMax << exponent;
+		allowedMax = allowedMax - 1;
+	}
+	return (min >= allowedMin && max <= allowedMax);
```
You could use something like `isIncludedIn(IntegerOverflowBounds::maximumRange(type))`

chriseth commented on libsolidity/analysis/IntegerOverflowBounds.cpp:
```diff
@@ -0,0 +1,267 @@
+	max = numerator(type.value());
+}
+
+void IntegerOverflowBounds::setToDefaultValue()
```
I think it would be cleaner if this class was immutable and these functions would be constructors or factory functions. What do you think?

roadriverrail replied:
I think you're probably on the right track.  Throughout Overflow checker, I'm making "blank" range objects, then pretty much immediately doing something to manipulate them.  After checking in the if/else code earlier today, I was starting to feel like maybe I needed to have functions that return range objects after manipulating them.

chriseth commented on libsolidity/analysis/IntegerOverflowBounds.cpp:
```diff
@@ -0,0 +1,267 @@
+	max = 0;
+}
+
+void IntegerOverflowBounds::constrain(OverflowBounds const& bounds)
```
Perhaps it would be clearer to just call it `intersect`?

roadriverrail replied:
Yes, with the proviso that I'm calling it `intersectBounds` to match `unionBounds`, because `union` is a reserved word.

chriseth commented on libsolidity/analysis/IntegerOverflowBounds.cpp:
```diff
@@ -0,0 +1,267 @@
+
+OverflowBounds& IntegerOverflowBounds::operator=(RationalNumberType const& number)
+{
+	min = numerator(number.value());
```
Please handle the case (assertion could be fine for now) where `denominator(number) != 1`

roadriverrail replied:
Will fix.

chriseth commented on libsolidity/analysis/OverflowChecker.cpp:
```diff
@@ -0,0 +1,220 @@
+		if (m_errorReporter.errors().empty())
+			throw; // Something is weird here, rather throw again.
+	}
+	return Error::containsOnlyWarnings(m_errorReporter.errors());
```
This component should only ever produce warnings, right? Perhaps it does not make sense to return a bool from this function anyway.

roadriverrail replied:
Good point.  Easy fix.

chriseth commented on libsolidity/analysis/OverflowChecker.cpp:
```diff
@@ -0,0 +1,220 @@
+	if (identifier)
+	{
+		shared_ptr<OverflowBounds>bounds = m_overflowInfo[&_assignment.rightHandSide()];
+		auto var = dynamic_cast<VariableDeclaration const*>(identifier->annotation().referencedDeclaration);
```
Please always assert after such a `dynamic_cast`.

roadriverrail replied:
This seems like a good place to ask the question.  I feel like I'm making a lot of `dynamic_casts` and I feel like it might be a sign of awkward code.  Here, it's obviously needed, but elsewhere, I'm often doing it just to force type compliance, and something feels wrong about that.

chriseth commented on libsolidity/analysis/OverflowChecker.cpp:
```diff
@@ -0,0 +1,220 @@
+		shared_ptr<OverflowBounds>bounds = m_overflowInfo[&_assignment.rightHandSide()];
+		auto var = dynamic_cast<VariableDeclaration const*>(identifier->annotation().referencedDeclaration);
+		m_varOverflowInfo[var] = bounds;
+		if (!bounds->fitsType(*(var->annotation().type)))
```
If there is such a more or less symmetric if-else, I would suggest to use the non-negated case first.

roadriverrail replied:
So, you would prefer:
```
                if (bounds->fitsType(*(var->annotation().type))) { /* do nothing */}
                else
                        m_errorReporter.warning(_assignment.location(), "Variable assignment may cause overflow.  " + bounds->toString());
```

chriseth commented on libsolidity/analysis/OverflowChecker.cpp:
```diff
@@ -0,0 +1,220 @@
+{
+	if (lockConstantExpression(_operation))
+		return false; // Do not descend
+	else
```
To reduce indentation, I would suggest to not use an explicit `else` case here.

roadriverrail replied:
Good point.

chriseth commented on libsolidity/analysis/OverflowChecker.cpp:
```diff
@@ -0,0 +1,220 @@
+		auto const* integerType = dynamic_cast<IntegerType const*>(_operation.annotation().commonType.get());
+		if (integerType)
+		{
+			//First, descend and resolve both sides of the operation.
```
I think we should do that in any case, regardless of the type. Consider:

`f(a+1)` - if `f` returns `void`, we still want to look deeper into the ast.

chriseth replied:
Actually it might make sense to use `endVisit` for the regular checks and only `visit` for the `lockConstantExpression` calls.

roadriverrail replied:
Right now, my code can't support function calls.  When I do get to that point, I will definitely *have* to evaluate the expressions in the parameters to a call because a call like `f(x++)` will generate a side-effect on `x`.  So, I will definitely descend all function calls.  In your example, when that descent reaches `a+1`, I wouldn't expect `lockConstantExpression( )` to filter it out because the expression `a+1` can't be a RationalNumberType.

roadriverrail replied:
Actually, I need to correct myself here.  Right now, in something like `f(a+1)`, we will check the `a+1` because it's a BinaryOperation.

chriseth commented on libsolidity/analysis/OverflowChecker.cpp:
```diff
@@ -0,0 +1,220 @@
+			bounds.performBinaryOp(_operation.getOperator(), *leftBounds, *rightBounds);
+			m_overflowInfo[&_operation] = make_shared<IntegerOverflowBounds>(bounds);
+			if (!bounds.fitsType(*integerType))
+				m_errorReporter.warning(_operation.location(), "Expression potentially overflowing its storage.  Range is: " + bounds.toString() + " Storage is: "+integerType->toString(false));
```
Please don`t use the word storage here, it has a special meaning in Solidity :-)

roadriverrail replied:
Will fix.

chriseth commented on libsolidity/analysis/OverflowChecker.cpp:
```diff
@@ -0,0 +1,220 @@
+					auto var = dynamic_cast<VariableDeclaration const*>(identifier->annotation().referencedDeclaration);
+					if (var)
+					{
+						IntegerOverflowBounds constraint(*_operation.leftExpression().annotation().type);
```
In the long run, the constraints should have different mechanics than the ranges as outlined in https://github.com/ethereum/solidity/issues/2274#issuecomment-310323184 - perhaps there should even be a different class that applies them.

roadriverrail replied:
I agree that constraints require a redesign, and I actually think that we'll ultimately need some sort of worker class that can generate them.  I'm just not 100% sure how I want that to work right now.

Really, ranges need a redesign, too, because they can't support being disjoint, but I'm not sure how I want that to work either (possibly with range objects being able to be composed of multiple ranges, too).

chriseth reviewed (commented):
Would be great to already have tests for cases that work to see the progress :-)

axic commented on libsolidity/analysis/IntegerOverflowBounds.cpp:
```diff
@@ -0,0 +1,267 @@
+		allowedMin = 0;
+		allowedMax = 1;
+		allowedMax = allowedMax << exponent;
+		allowedMax = allowedMax - 1;
```
Would these be easier to read as `(1 << exponent) - 1`?

axic replied:
Also this other change (https://github.com/ethereum/solidity/pull/2538#discussion_r126530260) could be utilized here.

roadriverrail replied:
I see that #2554 has been merged, so I can rebase to pick that up and then get rid of this redundant code.  Thanks for doing that!

axic commented on libsolidity/analysis/IntegerOverflowBounds.cpp:
```diff
@@ -0,0 +1,267 @@
+#include <libsolidity/analysis/IntegerOverflowBounds.h>
```
Missing license header.

axic replied:
Fixed.

axic commented on libsolidity/ast/ASTAnnotations.h:
```diff
@@ -23,6 +23,7 @@
 #pragma once
 
 #include <libsolidity/ast/ASTForward.h>
+#include <libsolidity/analysis/OverflowBounds.h>
```
I think this is not needed?

roadriverrail replied:
Will fix.

axic commented:
Please rebase.

roadriverrail reviewed (commented):
@chriseth @axic Thank you both for the attention to this change.  This feature still has a very long way to go, but I appreciate the feedback deeply so that I can fix a bunch of things as a I go.  I will be pushing a commit with your smaller corrections tonight.  I want to start looking at a refactor next so I have cleaner code at the top level.  I might take a break to write some tests first, though.

axic commented on libsolidity/ast/Types.cpp:
```diff
@@ -414,6 +414,14 @@ u256 IntegerType::literalValue(Literal const* _literal) const
 	return u256(_literal->value());
 }
 
+s256 IntegerType::signedValue(Literal const* _literal) const
```
This doesn't seem to be used yet?

I wonder if `rational value()` would be a better choice since it would be a unified interface between this and `RationalNumberType`?

roadriverrail commented:
Rebased and moved on to the helper functions for IntegerType.  Next on my list is to look to more of the design and refactoring remarks from @chriseth because the code is definitely becoming unwieldy to work with.

Other things I feel are coming up on the horizon that I need:

- Something to properly evaluate boolean expressions and determine the set of constraints that can be applied to ensure that it is true or false.  E.g. given the expression (x || y), it returns {{x}, {y}, {x,y}}.  This will be necessary for properly analyzing if/else, require( ), etc.

- Loops.  I have no good ideas about how to handle them.  The potential for overflow with a loop depends strongly on how many times the loop will run before terminating, and I don't have good intuitions about how to handle this.

That said, I do have some sense about how to do some refactoring to get this code cleaner and easier to read, and I also can work on extending this to analyze more than one function, and I can also work on expanding the mathematical operations supported, and so on.

axic commented on libsolidity/analysis/IntegerOverflowBounds.cpp:
```diff
@@ -0,0 +1,289 @@
+{
+	if (min > max)
+	{
+		dev::bigint swap = min;
```
Could this use boost's `swap` or is that overkill?

chriseth commented:
@roadriverrail I have not read all the recent changes, just commenting on your comment above:

I think a simple first version does not require the evaluation of all boolean expression perhaps apart from either simple `&&` or `||`, whichever is more convenient in that situation.

Concerning loops: What you can do is

 - analyze the loop about which local variables it might be affected
 - assume the loop condition inside the body to be true (just like an `if`), but clear all other assumptions about affected local variables
 - after the loop, clear all assumptions about affected local variables (even the loop codition).

roadriverrail commented:
@chriseth No worries about all the other comments.  I think your
refactoring to make the range objects immutable is a great idea, and I hope
to take it on either later today or tomorrow morning.  Can you explain what
you mean when you say "clear assumptions"?  If I "clear assumptions" from a
particular value range, am I basically returning it to [minval, maxval]?

On Tue, Jul 25, 2017 at 11:38 AM chriseth <notifications@github.com> wrote:

> @roadriverrail <https://github.com/roadriverrail> I have not read all the
> recent changes, just commenting on your comment above:
>
> I think a simple first version does not require the evaluation of all
> boolean expression perhaps apart from either simple && or ||, whichever
> is more convenient in that situation.
>
> Concerning loops: What you can do is
>
>    - analyze the loop about which local variables it might be affected
>    - assume the loop condition inside the body to be true (just like an if),
>    but clear all other assumptions about affected local variables
>    - after the loop, clear all assumptions about affected local variables
>    (even the loop codition).
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/ethereum/solidity/pull/2424#issuecomment-317776858>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAlhYChOkrCgf6qsS2TakjfHRUlnsLvdks5sRgwOgaJpZM4N-yhN>
> .
>

axic commented on libsolidity/analysis/IntegerOverflowBounds.cpp:
```diff
@@ -0,0 +1,254 @@
+	{
+		min = _right.getMin();
+		max = _right.getMax();
+	}
```
Should be an assert for the else here.

axic commented on libsolidity/analysis/IntegerOverflowBounds.cpp:
```diff
@@ -0,0 +1,254 @@
+	dev::bigint max;
+	// see here: https://en.wikipedia.org/wiki/Interval_arithmetic
+	// Not sure if modulus is right, but it feels right
+	if (_op == Token::Add)
```
Would a switch make more sense here?

axic commented on libsolidity/analysis/OverflowChecker.cpp:
```diff
@@ -0,0 +1,301 @@
+							invertOp = Token::Equal;
+							break;
+						default:
+							break;
```
Assert for this case?

Commit:
Gate overflow checking on pragma

This commit introduces the "analyzeIntegerOverflow" pragma to gate
the overflow checker.

Commit:
Add overflow checking pragma to tests

roadriverrail commented:
I've gone back to trying to consider loops again.  What I'm not really understanding in the explanation of loop analysis provided by @chriseth is how we determine when a loop will or will not produce an overflow.  It might be easier to work through an example of what I consider to be a short but logically challenging loop:

```
uint8 x = 0;
uint8 y = 0;
while (x < some_array.length) {
  y += x;
  x++;
}
```

This loop will overflow `y` for `some_array.length` > 22, but in order for me (personally) to calculate that, I had to recognize that `y` represents the triangular number sequence and to then apply the formula for triangular numbers for the value 255 and arrive at the result, 22, as being the largest triangular number <= 255.  I'm sure there's a way to algorithmically analyze this, but I don't know where to get started to even find out.  It requires recognizing the rates of change of `x` and `y` inside the loop and then working out how many times the loop can execute before it will overflow.  Determining rates of change that are constant across loop iterations (e.g. `x++`) is not so hard because you can simply compare the values at the beginning and end of the loop.  But it's determining rates of change that will vary with each loop iteration (e.g. `y += x` when `x` also changes in the loop) that I can't figure out.

There's also the question of how to determine if the logic of the loop will ensure it runs only a certain number of times.  I'm not clear on how, in a general case, we would be able to take "this loop is safe for 23 iterations" and apply it back to the constraints in the `while( )` expression.

It's possible I don't understand the scope envisioned here or that I don't know of a good method for solving this issue because it's outside my usual area of expertise.  @chriseth, what do you think?

axic commented:
Can you please rebase and use the new experimental feature pragma?

axic commented:
@chriseth @leonardoalt @ekpyron @bit-shift @christianparpart do you think it would be worth the effort finishing this PR or all we in favour of dropping this for the SMT option we have? The main pain point with SMT solvers we have is their code size and as a result we cannot bundle it in the compiler. This PR was started around the same time as the SMT prototype work and after that showed signs of success, the idea here was mostly to have a crude overflow detection system.

leonardoalt commented:
I lean toward the SMT option, assuming we manage to finish the JS work.

Related to that we have the runtime overflow checks we've been talking about, still without conclusive experimental results.

axic commented:
I tend to be on the same opinion that runtime overflow checks should provide the safeguard (and thus be detected during testing), while SMT would give the tool to help identify those earlier in development.

Downside is that if a runtime test is missed, the contract may end up in an unrecoverable state, but that is just a result of inadequate testing and auditing.

ekpyron commented:
Working towards making this PR reliable in all or at least most cases would probably mean reimplementing a custom SMT solver anyways, so I'd also lean towards dropping this in favor of the SMT option.
